monotonic/src/main/scala/com/criteo/cuttle/monotonic/MonotonicScheduler.scala

```diff
@@ -0,0 +1,165 @@
+package com.criteo.cuttle.monotonic
+
+import java.time.LocalDateTime
+import java.time.format.DateTimeFormatter
+
+import com.criteo.cuttle.{Executor, Job, Scheduler, Scheduling, SchedulingContext, Workflow, XA, utils}
+import doobie.imports._
+import io.circe.{Encoder, Json}
+
+import scala.concurrent.Future
+import scala.concurrent.duration.{Duration => ScalaDuration}
+import scala.concurrent.stm._
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.reflect.ClassTag
+
+object MonotonicContext {
+  def apply[T: Encoder: ClassTag](execParams: T, execTime: LocalDateTime) = new MonotonicContext[T] {
+    def executionParams = execParams
+
+    def executionTime = execTime
+
+    val enc = implicitly[Encoder[T]]
+
+    override def toJson: Json = Json.obj(
+      "exec_params" -> enc(execParams),
+      "exec_time" -> Json.fromString(DateTimeFormatter.ISO_LOCAL_DATE_TIME.format(executionTime))
+    )
+
+    override def log: ConnectionIO[String] = Database.serializeContext(this)
+
+    override def compareTo(other: SchedulingContext): Int = other match {
+      case o: MonotonicContext[T] => ordering.compare(this, o)
+    }
+  }
+
+  implicit def ordering[T] = Ordering.fromLessThan((t1: MonotonicContext[T], t2: MonotonicContext[T]) => t1.executionTime.isBefore(t2.executionTime))
+}
+
+trait MonotonicContext[T] extends SchedulingContext {
+  /**
+    * Parameters of jobs.
+    * Each graph execution will provide different values.
+    *
+    * @return
+    */
+  def executionParams: T
+
+  /**
+    * Workflow execution time, used to order executions.
+    * This is the time of the completion of the trigger.
+    *
+    * @return graph execution time
+    */
+  def executionTime: LocalDateTime
+
+  def enc: Encoder[T]
+}
+
+class TriggerJob[T](triggerName: String, trigger: () => Future[T], mon: MonotonicScheduling[T])
+  extends Job[MonotonicScheduling[T]](id = triggerName, name = "trigger_job", description = "Triggers the execution of a workflow", tags = Set.empty, scheduling = mon)(_ => Future.successful(())) {
+  def apply() = trigger()
+}
+
+case class MonotonicScheduling[T: ClassTag: Encoder](trigger: MonotonicScheduling.Aux[T]) extends Scheduling {
```

Comment:
It was a property of the scheduler but I had issues with typing and generics... That's why I moved everything to Scheduling. I'll try to solve the issues with typing and get back to you if I get stuck.

Users don't handle any loop, it's handled by the code. But yes they still need to provide a function that need to handle failures and retries (for example: I trigger the execution of a workflow when data is retrieved from an API, if there's no data I continue to query the API).

I'll see how to make it "mandatory" and maybe provide a default implementation (a timer).